src/qibolab/_core/instruments/qcs.py

```diff
@@ -0,0 +1,343 @@
+"""Qibolab driver for Keysight QCS instrument set."""
+
+from typing import Optional, Union
+
+import keysight.qcs as qcs  # pylint: disable=E0401
+import numpy as np
+
+from qibolab._core.components import AcquisitionChannel, Config, DcChannel, IqChannel
+from qibolab._core.execution_parameters import (
+    AcquisitionType,
+    AveragingMode,
+    ExecutionParameters,
+)
+from qibolab._core.identifier import ChannelId, Result
+from qibolab._core.instruments.abstract import Controller
+from qibolab._core.pulses import Drag, Envelope, Gaussian, PulseId, Rectangular
+from qibolab._core.sequence import InputOps, PulseSequence
+from qibolab._core.sweeper import ParallelSweepers, Parameter
+from qibolab._core.unrolling import Bounds
+
+NANOSECONDS = 1e-9
+
+BOUNDS = Bounds(
+    waveforms=1,
+    readout=1,
+    instructions=1,
+)
+
+__all__ = ["KeysightQCS"]
+
+
+def generate_qcs_envelope(shape: Envelope) -> qcs.Envelope:
+    """Converts a Qibolab pulse envelope to a QCS Envelope object."""
+    if isinstance(shape, Rectangular):
+        return qcs.ConstantEnvelope()
+
+    elif isinstance(shape, (Gaussian, Drag)):
+        return qcs.GaussianEnvelope(shape.rel_sigma)
+
+    else:
+        # TODO: Rework this code to support other Qibolab pulse envelopes
+        # raw_envelope = shape.i(num_samples) + 1j * shape.q(num_samples)
+        # return qcs.ArbitraryEnvelope(
+        #    times=np.linspace(0, 1, num_samples), amplitudes=raw_envelope
+        # )
+        raise Exception("Envelope not supported")
+
+
+def generate_qcs_rfwaveform(
+    duration: Union[float, qcs.Scalar],
+    envelope: Envelope,
+    amplitude: Union[float, qcs.Scalar],
+    frequency: Union[float, qcs.Scalar],
+    phase: Union[float, qcs.Scalar],
+) -> qcs.RFWaveform:
+
+    return qcs.RFWaveform(
+        duration=duration,
+        envelope=generate_qcs_envelope(envelope),
+        amplitude=amplitude,
+        rf_frequency=frequency,
+        instantaneous_phase=phase,
+    )
+
+
+class KeysightQCS(Controller):
+    """Driver for interacting with QCS controller server."""
+
+    bounds: str = "qcs/bounds"
+
+    # Map of QCS virtual channels to QCS physical channels
+    qcs_channel_map: qcs.ChannelMapper
+    # Map of Qibolab channel IDs to QCS virtual channels
+    virtual_channel_map: dict[ChannelId, qcs.Channels]
+    # Map of QCS virtual acquisition channels to QCS state classifiers
+    classifier_map: Optional[dict[qcs.Channels, qcs.MinimumDistanceClassifier]] = None
+
+    def connect(self):
+        self.backend = qcs.HclBackend(self.qcs_channel_map, hw_demod=True)
+        self.backend.is_system_ready()
+
+    @property
+    def sampling_rate(self):
+        return qcs.SAMPLE_RATES[qcs.InstrumentEnum.M5300AWG] / 1e9
```

Comment:
Oh, I was basing it on QM's
https://github.com/qiboteam/qibolab/blob/main/src/qibolab/_core/instruments/qm/controller.py#L200-L203

Reply:
Correct, I could have pointed out even there.

@stavros11

Reply:
This is defined as an `abstractmethod` at 
https://github.com/qiboteam/qibolab/blob/main/src/qibolab/_core/instruments/abstract.py#L62-L66
So I can't implement it as a module-level constant

Reply:
Ah, yes, it is actually used in the `Platform`:
https://github.com/qiboteam/qibolab/blob/946a04f21f522cd05f898de91270153af7cdceff/src/qibolab/_core/platform/platform.py#L139
(what for, not sure, but it's true that is there - and it will be there until the next major release - at most we could open an issue and move towards deprecation)

Reply:
Fixed in https://github.com/qiboteam/qibolab/pull/944/commits/3e34eb68477500835cd059a0f9158cefd25ee224